[Fix] Renaming or removing a contracted model raises a BreakingChange warning/error

MichelleArk opened the pull request:
resolves https://github.com/dbt-labs/dbt-core/issues/10116

<!---
  Include the number of the issue addressed by this PR above if applicable.
  PRs for code changes without an associated issue *will not be merged*.
  See CONTRIBUTING.md for more information.

  Add the `user docs` label to this PR if it will need docs changes.  An 
  issue will get opened in docs.getdbt.com upon successful merge of this PR.
-->

### Problem

<!---
  Describe the problem this PR is solving. What is the application state
  before this PR is merged?
-->

Deleting a contracted model does not warn or error regarding a breaking contract change

### Solution
* detect deleted nodes in StateModifiedSelector
* add same_contract_deleted to ModelNode to handle deletion path

<!---
  Describe the way this PR solves the above problem. Add as much detail as you
  can to help reviewers understand your changes. Include any alternatives and
  tradeoffs you considered.
-->

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me  
- [x] I have run this code in development and it appears to resolve the stated issue  
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc) or this PR has already received feedback and approval from Product or DX
- [x] This PR includes [type annotations](https://docs.python.org/3/library/typing.html) for new and modified functions

MichelleArk commented on core/dbt/graph/selector_methods.py:
```diff
@@ -719,7 +719,9 @@ def check_modified_contract(
         # get a function that compares two selector target based on compare method provided
         def check_modified_contract(old: Optional[SelectorTarget], new: SelectorTarget) -> bool:
-            if hasattr(new, compare_method):
+            if new is None and hasattr(old, compare_method + "_deleted"):
```
can't say I'm in love with this it was the best way forward I could think of that: 
1. reuses the `state:modified.contract` user-facing method ([configured here](https://github.com/dbt-labs/dbt-core/pull/10221/files#diff-9aa4de7075574ef36d838665148e5858294fae6fd66c0eed1fcf2033a34ee84eR752)) to detect both changes that affect deleted + non-deleted changes. 
2. keep the diff method localized to the node type instead of a more global search logic in the StateSelection method which felt like overloading what it did + deviating from the current pattern too much.

open to other suggestions!

jtcohen6 commented on core/dbt/graph/selector_methods.py:
```diff
@@ -785,6 +787,16 @@ def search(self, included_nodes: Set[UniqueId], selector: str) -> Iterator[Uniqu
                 yield unique_id
 
+        # checkers that can handle deleted nodes
+        if checker.__name__ in ["check_modified_contract"]:
```
✅ m8

gshank reviewed (approved):
Looks good. Is this something that applies to disabled nodes too, or is that handled somewhere else?

MichelleArk commented:
> Is this something that applies to disabled nodes too, or is that handled somewhere else?

hmm, good question. I don't believe we should be raising breaking contract changes for disabled nodes that have been deleted/renamed but cc'ing @jtcohen6 to confirm. Either way it'd be good to add a test around that.

jtcohen6 commented:
> > Is this something that applies to disabled nodes too, or is that handled somewhere else?
> 
> hmm, good question. I don't believe we should be raising breaking contract changes for disabled nodes that have been deleted/renamed but cc'ing @jtcohen6 to confirm. Either way it'd be good to add a test around that.

@gshank @MichelleArk Good catch, thank you for thinking of this!
- If a node was previously disabled, it is not a breaking change to remove/rename it
- If a node was previously enabled, with an enforced contract, and it is now _disabled_, I do believe that is a breaking change. (Same as if the node were deleted - the effect is that it's removed from `manifest.nodes`)

MichelleArk commented:
> If a node was previously enabled, with an enforced contract, and it is now disabled, I do believe that is a breaking change. (Same as if the node were deleted - the effect is that it's removed from manifest.nodes)

Thank you! That's a case worth handling in this PR.

Updated the "same_contract_deleted" method to "same_contract_removed" and plumbed disabled nodes through the state method selection method and added testing. Should be good for another look @gshank 👍

gshank commented:
Looks good!  It's nice that adding disabled nodes was just a minor code change (and another test).
